docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,100 @@
+
+## Why
+
+The single biggest source of frustration from internal Red Hat users of Observatorium right now is the time it takes for changes to their recording rules to appear in our production infrastructure.
+
+### Pitfalls of the current solution
+
+#### Implicit deploy cycle dependency
+
+Currently, the only method of modifying Prometheus recording rules in Observatorium is via the jsonnet definition in our repository. This implicitly ties an update of the rule configuration to a rollout of the entire infrastructure. This implicitly ties the deploy-cycle of Prometheus rules to the deploy-cycle of Observatorium itself. If a new set of rules needs to be deployed, our team is required to roll-out our production infrastructure.
+
+As the number of tenants we support and users we serve increases, the team responsible for the Observatorium installation is required to roll out production more frequently in order to satisfy user requests. This becomes an increasing impediment to user experience as the size of Observatorium increases.
+
+#### Shared configuration repositories
+
+For rules to be rolled out, users need the ability to raise pull requests against the Observatorium repositories that contain our source configuration. This works fine when the two teams co-exist within the same organisation, but it becomes a blocking issue when the two teams share a much more restricted (e.g. inter-company) trust boundary.
+
+This is not a blocker for Red Hat's internal offering, however this is not inline with Observatorium's stated goal (TODO: insert link) of offering a SaaS-like monitoring solution.
+
+## Goals
+
+* Users can update their Prometheus recording rules without intervention from the operating team.
+* The Observatorium API must be able to scale horizontally scale with the number of users.
+
+## Non-Goals
+
+* Rate-limiting or accounting of user resources.
+* Define or implement the underlying storage technology.
+
+## How
+
+We propose to solve the above problem by implementing a multi-tenant API that allows users to create, read, update and delete Prometheus alerting and recording rules.
+
+Thanos Ruler operates on the rule configuration provided by users, which is consumed from a file defined by Ruler's `--rule.file` flag. Therefore, this configuration that we ingest via the API needs to be replicated into the local data directory of the Thanos Ruler instances. This is achieved with the [thanos-ruler-syncer](https://github.com/observatorium/thanos-rule-syncer), an application that will call the API above and write the results into a location that Thanos Rule can access.
+
+### How should we persist user's rule configuration?
+
+For the Observatorium API to satisfy our requirement of scaling horizontally with the number of user requests, we need to use a persistent storage layer for our user's recording rules.
+
+TODO
+* Object storage is already used in Thanos. Users will already have this infrastructure provisioned.
```

Comment:
@squat I'll need a hand in filling out these sections - happy to do either over meet or happy for you to fill this in async